spring-cloud-gcp-data-firestore/src/main/java/org/springframework/cloud/gcp/data/firestore/FirestoreTemplate.java

```diff
@@ -400,34 +411,56 @@ private void doIfTransaction(Context ctx, Consumer<ReactiveFirestoreResourceHold
 	}
 
 	private <T> Write createUpdateWrite(T entity) {
-		String documentResourceName = buildResourceName(entity);
-		Document document = getClassMapper().entityToDocument(entity, documentResourceName);
-		return Write.newBuilder()
-				.setUpdate(document)
-				.build();
+		Builder builder = Write.newBuilder();
+		if (getIdValue(entity) == null) {
+			builder.setCurrentDocument(Precondition.newBuilder().setExists(false).build());
+		}
+		String resourceName = buildResourceName(entity);
+		Document document = getClassMapper().entityToDocument(entity, resourceName);
+		return builder.setUpdate(document).build();
 	}
 
 	private <T> String buildResourceName(T entity) {
 		FirestorePersistentEntity<?> persistentEntity =
 				this.mappingContext.getPersistentEntity(entity.getClass());
 		FirestorePersistentProperty idProperty = persistentEntity.getIdPropertyOrFail();
 		Object idVal = persistentEntity.getPropertyAccessor(entity).getProperty(idProperty);
+		if (idVal == null) {
+			if (idProperty.getType() != String.class) {
+				throw new FirestoreDataException("Automatic ID generation only supported for String type");
+			}
 
+			//TODO: replace with com.google.cloud.firestore.Internal.autoId() when it is available
+			idVal = autoId();
+			persistentEntity.getPropertyAccessor(entity).setProperty(idProperty, idVal);
+		}
 		return buildResourceName(persistentEntity, idVal.toString());
 	}
 
 	private String buildResourceName(FirestorePersistentEntity<?> persistentEntity, String s) {
 		return this.parent + "/" + persistentEntity.collectionName() + "/" + s;
 	}
 
-	private String getIdValue(Object entity, FirestorePersistentEntity persistentEntity) {
+	private Object getIdValue(Object entity) {
+		FirestorePersistentEntity<?> persistentEntity =
+						this.mappingContext.getPersistentEntity(entity.getClass());
 		FirestorePersistentProperty idProperty = persistentEntity.getIdPropertyOrFail();
-		Object idVal = persistentEntity.getPropertyAccessor(entity).getProperty(idProperty);
 
-		return idVal.toString();
+		return persistentEntity.getPropertyAccessor(entity).getProperty(idProperty);
 	}
 
 	public FirestoreClassMapper getClassMapper() {
 		return this.classMapper;
 	}
+
+	//Creates a pseudo-random 20-character ID that can be used for Firestore documents.
+	//Copied from com.google.cloud.firestore.FirestoreImpl
+	private static String autoId() {
```

Comment:
Could you move the Id generating helper method `autoId()` to a package-private class? maybe something like `FirestoreIdGenerator`.

This way you can move all the code and fields in there related to ID generation; which allows you to remember what to delete once `com.google.cloud.firestore.Internal.autoId()` becomes available. Otherwise you will have to comb through the code a little bit more in the future to do this removal.